Sources/third_party/Experimental/Complex.swift

```diff
@@ -0,0 +1,325 @@
+        guard let t = T(exactly: source) else { return nil }
+        self.real = t
+        self.imaginary = 0
+    }
+
+    @differentiable(vjp: _vjpMultiply(lhs:rhs:) where T : Differentiable)
+    static func * (lhs: Complex, rhs: Complex) -> Complex {
+        var a = lhs.real, b = lhs.imaginary, c = rhs.real, d = rhs.imaginary
+        let ac = a * c, bd = b * d, ad = a * d, bc = b * c
+        let x = ac - bd
+        let y = ad + bc
+
+        if x.isNaN && y.isNaN {
+            var recalculate = false
+            if a.isInfinite || b.isInfinite {
+                a = T(signOf: a, magnitudeOf: a.isInfinite ? 1 : 0)
+                b = T(signOf: b, magnitudeOf: b.isInfinite ? 1 : 0)
+                if c.isNaN { c = T(signOf: c, magnitudeOf: 0) }
+                if d.isNaN { d = T(signOf: d, magnitudeOf: 0) }
+                recalculate = true
+            }
+            if c.isInfinite || d.isInfinite {
+                if a.isNaN { a = T(signOf: a, magnitudeOf: 0) }
+                if b.isNaN { b = T(signOf: b, magnitudeOf: 0) }
+                c = T(signOf: c, magnitudeOf: c.isInfinite ? 1 : 0)
+                d = T(signOf: d, magnitudeOf: d.isInfinite ? 1 : 0)
+                recalculate = true
+            }
+            if !recalculate &&
+                (ac.isInfinite || bd.isInfinite || ad.isInfinite || bc.isInfinite) {
+                if a.isNaN { a = T(signOf: a, magnitudeOf: 0) }
+                if b.isNaN { b = T(signOf: b, magnitudeOf: 0) }
+                if c.isNaN { c = T(signOf: c, magnitudeOf: 0) }
+                if d.isNaN { d = T(signOf: d, magnitudeOf: 0) }
+                recalculate = true
+            }
+            if recalculate {
+                return Complex(
+                real: .infinity * (a * c - b * d),
+                imaginary: .infinity * (a * d + b * c)
+                )
+            }
+        }
+        return Complex(real: x, imaginary: y)
+    }
+
+    static func *= (lhs: inout Complex, rhs: Complex) {
+        lhs = lhs * rhs
+    }
+
+    var magnitude: T {
+        var x = abs(real)
+        var y = abs(imaginary)
+        if x.isInfinite { return x }
+        if y.isInfinite { return y }
+        if x == 0 { return y }
+        if x < y { swap(&x, &y) }
+        let ratio = y / x
+        return x * (1 + ratio * ratio).squareRoot()
+    }
+}
+
+extension Complex : SignedNumeric {
+    @differentiable(vjp: _vjpNegate where T : Differentiable)
+    static prefix func - (operand: Complex) -> Complex {
+        return Complex(real: -operand.real, imaginary: -operand.imaginary)
+    }
+
+    mutating func negate() {
+        real.negate()
+        imaginary.negate()
+    }
+}
+
+extension Complex {
+    @differentiable(vjp: _vjpDivide(lhs:rhs:) where T : Differentiable)
+    static func / (lhs: Complex, rhs: Complex) -> Complex {
+        var a = lhs.real, b = lhs.imaginary, c = rhs.real, d = rhs.imaginary
+        var x: T
```

Comment:
In the conditionals below, since assignments to `x` and `y` on all paths of execution are initialization instead of mutation, you can declare `x` and `y` as `let`.

Reply:
Let me try, but I don't think it will as we have:
```swift
if c.magnitude >= d.magnitude {
    x = (a + b * ratio) / denominator
    y = (b - a * ratio) / denominator
} else {
    x = (a * ratio + b) / denominator
    y = (b * ratio - a) / denominator
}
if x.isNaN && y.isNaN {
  // More if logic
  x = T(signOf: c, magnitudeOf: .infinity) * a
  y = T(signOf: c, magnitudeOf: .infinity) * b
}
```
So there is a path of execution that may set it again, specifically when we doing something equivalent to `0.0/0.0` which would be nan

Reply:
Oh you're right, I missed it!